File: face/face.py
import sys
import random
from PyQt5.QtWidgets import QWidget
from PyQt5.QtGui import QPainter, QColor, QBrush, QPainterPath
from PyQt5.QtCore import Qt, QTimer, pyqtSignal

class Prototype_1FaceUI(QWidget):
    update_gaze_signal = pyqtSignal(float, float) 
    update_emotion_signal = pyqtSignal(str)       

    def __init__(self):
        super().__init__()
        self.setWindowTitle("Prototype_1 Eye - Face UI")
        
        self.resize(800, 480) 
        self.setStyleSheet("background-color: black;")

        self.current_gaze_x = 0.0
        self.current_gaze_y = 0.0
        self.target_gaze_x = 0.0
        self.target_gaze_y = 0.0
        
        self.max_gaze_offset = 300

        self.top_cut = 0.0
        self.bot_cut = 0.0
        
        self.emotion_target_top = 0.0
        self.emotion_target_bot = 0.0
        self.is_blinking = False

        self.update_gaze_signal.connect(self.set_gaze)
        self.update_emotion_signal.connect(self.set_emotion)

        self.timer = QTimer()
        self.timer.timeout.connect(self.update_physics)
        self.timer.start(16)
        
        self.blink_timer = QTimer()
        self.blink_timer.timeout.connect(self.start_blink)
        self.blink_timer.start(random.randint(2000, 6000))

    def start_blink(self):
        self.is_blinking = True
        QTimer.singleShot(150, self.end_blink)
        self.blink_timer.setInterval(random.randint(2000, 6000))

    def end_blink(self):
        self.is_blinking = False

    def set_gaze(self, x, y):
        self.target_gaze_x = max(-1.0, min(1.0, x * 1.3))
        self.target_gaze_y = max(-1.0, min(1.0, y * 1.3))

    def set_emotion(self, emotion):
        if emotion == "normal":
            self.emotion_target_top = 0
            self.emotion_target_bot = 0
        elif emotion == "happy":
            self.emotion_target_top = 0
            self.emotion_target_bot = 40
        elif emotion == "sad":
            self.emotion_target_top = 40
            self.emotion_target_bot = 0

    def update_physics(self):
        gaze_smoothness = 0.1 
        self.current_gaze_x += (self.target_gaze_x - self.current_gaze_x) * gaze_smoothness
        self.current_gaze_y += (self.target_gaze_y - self.current_gaze_y) * gaze_smoothness
        
        if self.is_blinking:
            target_top = 280
            target_bot = 0
            cut_smoothness = 0.6
        else:
            target_top = self.emotion_target_top
            target_bot = self.emotion_target_bot
            cut_smoothness = 0.15

        self.top_cut += (target_top - self.top_cut) * cut_smoothness
        self.bot_cut += (target_bot - self.bot_cut) * cut_smoothness
        
        self.update()

    def paintEvent(self, event):
        painter = QPainter(self)
        painter.setRenderHint(QPainter.Antialiasing, True)
        painter.fillRect(self.rect(), Qt.black)

        eye_w = 160
        eye_h = 280
        eye_spacing = 300
        center_x = self.width() / 2
        center_y = self.height() / 2
        Prototype_1_blue = QColor("#42a5f5")
        
        gaze_offset_x = self.current_gaze_x * self.max_gaze_offset
        gaze_offset_y = self.current_gaze_y * self.max_gaze_offset

        self.draw_true_geometry_eye(
            painter, cx=center_x - eye_spacing/2 + gaze_offset_x, cy=center_y + gaze_offset_y, 
            w=eye_w, h=eye_h, color=Prototype_1_blue
        )
        self.draw_true_geometry_eye(
            painter, cx=center_x + eye_spacing/2 + gaze_offset_x, cy=center_y + gaze_offset_y, 
            w=eye_w, h=eye_h, color=Prototype_1_blue
        )

    def draw_true_geometry_eye(self, painter, cx, cy, w, h, color):
        painter.save()
        painter.translate(cx, cy)
        
        base_path = QPainterPath()
        base_path.addRoundedRect(-w/2, -h/2, w, h, 80, 80)
        
        visible_y = -h/2 + self.top_cut
        visible_h = max(0, h - self.top_cut - self.bot_cut)
        
        cut_box = QPainterPath()
        cut_box.addRect(-w, visible_y, w*2, visible_h)
        
        final_eye_path = base_path.intersected(cut_box)
        
        painter.setBrush(QBrush(color))
        painter.setPen(Qt.NoPen)
        painter.drawPath(final_eye_path)
        painter.restore()

File: frontend/face.js
// ==========================================
// PROTOTYPE WEB - FACE UI (HTML5 CANVAS)
// ==========================================

class FaceUI {
    constructor() {
        console.log("[FACE UI] Initializing Canvas UI...");
        
        // Setup Canvas
        this.canvas = document.createElement('canvas');
        this.ctx = this.canvas.getContext('2d');
        this.canvas.style.display = 'block';
        this.canvas.style.width = '100vw';
        this.canvas.style.height = '100vh';
        this.canvas.style.backgroundColor = 'black';
        document.body.appendChild(this.canvas);
        
        // Remove margins/padding on body to ensure full screen
        document.body.style.margin = '0';
        document.body.style.overflow = 'hidden';

        // Physics Variables
        this.currentGazeX = 0.0;
        this.currentGazeY = 0.0;
        this.targetGazeX = 0.0;
        this.targetGazeY = 0.0;
        this.maxGazeOffset = 300;

        // Eyelid Cuts
        this.topCut = 0.0;
        this.botCut = 0.0;
        this.emotionTargetTop = 0.0;
        this.emotionTargetBot = 0.0;
        this.isBlinking = false;
        
        // Core Colors
        this.prototypeBlue = "#42a5f5";

        // Bind resize event
        window.addEventListener('resize', () => this.resizeCanvas());
        
        this.init();
    }

    init() {
        this.resizeCanvas();
        this.scheduleBlink();
        this.renderLoop();
    }

    resizeCanvas() {
        // Handle high-DPI displays (Retina screens) for crisp rendering
        const dpr = window.devicePixelRatio || 1;
        this.canvas.width = window.innerWidth * dpr;
        this.canvas.height = window.innerHeight * dpr;
        this.ctx.scale(dpr, dpr);
    }

    // --- STATE SETTERS ---
    setGaze(x, y) {
        // Map Gaze coordinates (-1.0 to 1.0), amplifying slightly
        this.targetGazeX = Math.max(-1.0, Math.min(1.0, x * 1.3));
        this.targetGazeY = Math.max(-1.0, Math.min(1.0, y * 1.3));
    }

    setEmotion(emotion) {
        if (emotion === "normal") {
            this.emotionTargetTop = 0;
            this.emotionTargetBot = 0;
        } else if (emotion === "happy") {
            this.emotionTargetTop = 0;
            this.emotionTargetBot = 40;
        } else if (emotion === "sad") {
            this.emotionTargetTop = 40;
            this.emotionTargetBot = 0;
        }
    }

    // --- BLINK LOGIC ---
    scheduleBlink() {
        // Random interval between 2s and 6s
        const nextBlink = Math.floor(Math.random() * (6000 - 2000 + 1) + 2000);
        setTimeout(() => this.startBlink(), nextBlink);
    }

    startBlink() {
        this.isBlinking = true;
        // Blink duration is 150ms
        setTimeout(() => {
            this.isBlinking = false;
            this.scheduleBlink();
        }, 150);
    }

    // --- PHYSICS & RENDERING ---
    updatePhysics() {
        // 1. Smooth Gaze Movement (Linear Interpolation)
        const gazeSmoothness = 0.1;
        this.currentGazeX += (this.targetGazeX - this.currentGazeX) * gazeSmoothness;
        this.currentGazeY += (this.targetGazeY - this.currentGazeY) * gazeSmoothness;

        // 2. Smooth Eyelid Movement
        let targetTop, targetBot, cutSmoothness;
        
        if (this.isBlinking) {
            targetTop = 280; // Full height of eye
            targetBot = 0;
            cutSmoothness = 0.6; // Fast blink down
        } else {
            targetTop = this.emotionTargetTop;
            targetBot = this.emotionTargetBot;
            cutSmoothness = 0.15; // Smooth emotion transition
        }

        this.topCut += (targetTop - this.topCut) * cutSmoothness;
        this.botCut += (targetBot - this.botCut) * cutSmoothness;
    }

    drawTrueGeometryEye(cx, cy, w, h, color) {
        this.ctx.save();
        this.ctx.translate(cx, cy);

        // Calculate visible bounding box after clipping
        const visibleY = -h / 2 + this.topCut;
        const visibleH = Math.max(0, h - this.topCut - this.botCut);

        // Setup Eyelid Clipping Region (Mimics QPainterPath.intersected)
        this.ctx.beginPath();
        this.ctx.rect(-w, visibleY, w * 2, visibleH);
        this.ctx.clip(); // Everything drawn after this will be constrained to this box

        // Draw the base rounded rectangle eye
        this.ctx.beginPath();
        // ctx.roundRect(x, y, width, height, radii) - Modern Canvas API
        this.ctx.roundRect(-w / 2, -h / 2, w, h, 80);
        
        this.ctx.fillStyle = color;
        this.ctx.fill();

        this.ctx.restore();
    }

    renderLoop() {
        this.updatePhysics();

        // Clear previous frame
        this.ctx.fillStyle = 'black';
        this.ctx.fillRect(0, 0, window.innerWidth, window.innerHeight);

        // Core Dimensions (Matching face.py)
        const eyeW = 160;
        const eyeH = 280;
        const eyeSpacing = 300;
        const centerX = window.innerWidth / 2;
        const centerY = window.innerHeight / 2;
        
        // Calculate Gaze Offsets
        const gazeOffsetX = this.currentGazeX * this.maxGazeOffset;
        const gazeOffsetY = this.currentGazeY * this.maxGazeOffset;

        // Draw Left Eye
        this.drawTrueGeometryEye(
            centerX - eyeSpacing / 2 + gazeOffsetX, 
            centerY + gazeOffsetY, 
            eyeW, 
            eyeH, 
            this.prototypeBlue
        );

        // Draw Right Eye
        this.drawTrueGeometryEye(
            centerX + eyeSpacing / 2 + gazeOffsetX, 
            centerY + gazeOffsetY, 
            eyeW, 
            eyeH, 
            this.prototypeBlue
        );

        // Queue next frame
        requestAnimationFrame(() => this.renderLoop());
    }
}

// Export to global window object
window.addEventListener('load', () => {
    window.FaceUI = new FaceUI();
});